numpy/lib/shape_base.py

```diff
@@ -1159,23 +1159,19 @@ def kron(a, b):
     bs = (1,)*max(0, nda-ndb) + bs
 
     # Compute the product
-    a_arr = _nx.asarray(a).reshape(a.size, 1)
-    b_arr = _nx.asarray(b).reshape(1, b.size)
-    result = a_arr * b_arr
+    a_arr = a.reshape(a.size, 1)
+    b_arr = b.reshape(1, b.size)
+    is_any_mat = isinstance(a_arr, matrix) or isinstance(b_arr, matrix)
+    # In case of `mat`, convert result to `array`
+    result = _nx.multiply(a_arr, b_arr, subok=(not is_any_mat))
 
     # Reshape back
     result = result.reshape(as_+bs)
     transposer = _nx.arange(nd*2).reshape([2, nd]).ravel(order='f')
     result = result.transpose(transposer)
     result = result.reshape(_nx.multiply(as_, bs))
 
-    wrapper = get_array_prepare(a, b)
-    if wrapper is not None:
-        result = wrapper(result)
-    wrapper = get_array_wrap(a, b)
-    if wrapper is not None:
-        result = wrapper(result)
-    return result
+    return result if not is_any_mat else matrix(result)
```

Comment:
Maybe `matrix(result, copy=False)` would help performance (not that I think it matters too much)

Reply:
Yeah, lets not worry about matrix performance.  Do matrix subclasses matter?  If we want to be on the safe side, I would otherwise suggest to keep using the array wrap logic in that branch (not the array-prepare part, I don't really think it makes any sense).
OTOH, matrix is already a niche use-case, so subclasses of matrix probably just don't matter.

Reply:
Wow that really made all the difference
```
BENCHMARKS NOT SIGNIFICANTLY CHANGED.
```

Reply:
> matrix is already a niche use-case, so subclasses of matrix probably just don't matter.

Yeah need to see if we can generalize the solution in the future instead of handling like what's done here.